### ISSUE TITLE
"Number.isinteger()" javascript error of Slider in Internet Explorer

### ISSUE THREAD

**Issue by gepcel** [bokeh/bokeh#7373]
I've an bokeh app runs by `bokeh serve app`.  It works under Chrome, and Edge, but not Internet Explorer. Do anybody know why?

In chrome or edge, the plot is like this:
<p><image src="https://user-images.githubusercontent.com/4179106/34599026-599f20b2-f22b-11e7-8163-2fa636570c27.png" width=400/></p>

But in Internet Explorer, it is like (no points showed up):
<p><image src="https://user-images.githubusercontent.com/4179106/34599054-7115be04-f22b-11e7-861f-06c53a436bfa.png" width=400/></p>

And the console keeps getting error messages like:
```cmd
2018-01-05 15:13:12,991 received invalid integer in pong b''
Traceback (most recent call last):
  File "C:\Anaconda3\lib\site-packages\bokeh\server\views\ws.py", line 161, in on_pong
    self.latest_pong = int(codecs.decode(data, 'utf-8'))
ValueError: invalid literal for int() with base 10: ''
2018-01-05 15:13:12,995 received invalid integer in pong b''
Traceback (most recent call last):
  File "C:\Anaconda3\lib\site-packages\bokeh\server\views\ws.py", line 161, in on_pong
    self.latest_pong = int(codecs.decode(data, 'utf-8'))
ValueError: invalid literal for int() with base 10: ''
2018-01-05 15:13:50,016 received invalid integer in pong b''
Traceback (most recent call last):
  File "C:\Anaconda3\lib\site-packages\bokeh\server\views\ws.py", line 161, in on_pong
    self.latest_pong = int(codecs.decode(data, 'utf-8'))
ValueError: invalid literal for int() with base 10: ''
```

Anybody knows why?

System and versions:
   
    python: 3.6.2
    bokeh: 0.12.10
    os: Windows 10 x64 Version 1709 - Chinese Language
    Internet Explorer: 11 (11.125.16299.0)
    tornado: 4.5.2

## Later Update:

With help from many kind people, there seems to be two problems. The first one is about a javascript method `Number.isinteger()` not supported in IE. The second one is about an inconsistent lengths error. Since @mattpap has tagged the first one a bug, I think it's better to make this issue about error 1. I'll open another issue with an minimum reproducable example.

**Comment by bryevdv** [comment]
Can you provide a minimal reproducing example that we can run and investigate?

**Comment by bryevdv** [comment]
Also, the `ping` messages are unrelated. Some browsers don't implement ping to spec. There's nothing we can do about that. but it's also harmless.

**Comment by mattpap** [comment]
Also output from IE's JavaScript console would be useful, as well as having this reproduced for most recent version of bokeh (0.12.13).

**Comment by gepcel** [comment]
First of all, thanks @mattpap, but bokeh v0.12.13 doesn't work. 

My app is a little bit complicated to upload as an example. Anyway, I've found that there are two problems. And I provided an example (at the ends) to the second problem and some describe to the first one.

## Problem 1

In my app, I used some dateframes, with 36 rows, and I use `ColumnDataSource(dict(latitude=df.y.values, ...)` as the y values  of points.  And I've found these:

* In Chrome or Edge or Firefox, all works perfectly, nothing wrong.
* But in IE:
    * The console gives an error like `[bokeh] data source has columns of inconsistent lengths`, 
    * And the `cmd console` gives an error as `c:\anaconda3\lib\site-packages\bokeh\models\sources.py:137: BokehUserWarning: ColumnDataSource's columns must be of the same length. Current lengths: ('date', 36), ('display_value', 36), ('fillcolor', 36), ('latitude', 72), ('longitude', 36), ('point_size', 36)`. 
    * Note that all datas including latitude (df['y'] have the same length of 36, not 72.
    * Change it to `ColumnDataSource(dict(latitude=df.y.tolist(), ...)`, solved this problem.
    * Can't think of a reason of difference between  `.values`  and `.tolist()`. Both works in Chrome, only `.tolist()` works in IE.
    * Can't be sure if or not there's a bug in my code. I'll keep thinking and testing.

## Problem 2

For this, I've managed to provide a minimal example (codes at the end). Generate 4 dataframes, with `x`, `y` of points, and `z` as points sizes. But:

1. In Chrome, it works. (Run by `bokeh serve --show test.py`)
2. In IE, the initialization plot is successful. The plot shows up.
3. But, the slider doesn't work. The IE console gives an error like "[SCRIPT438](http://msdn.microsoft.com/query/dev12.query?appId=Dev12IDEF1&l=ZH-CN&k=k(VS.WebClient.Help.SCRIPT438)): 对象不支持“isInteger'属性或方法'. Translate it to English would be something like "SCRIPT438: Object doesn't support 'isInteger attribute or method'" 

**Example**: 
```python
from bokeh.layouts import column, row
from bokeh.models.widgets import Slider
from bokeh.io import curdoc
import numpy as np
import pandas as pd
from bokeh.models import ColumnDataSource
from bokeh.plotting import figure

# Prepare the data
random = np.random.RandomState(0)
N = 4
datas = [pd.DataFrame(random.randint(1, 10, size=(10, 3)),
                      columns=['x', 'y', 'z']) for i in range(N)]

# Initialize the plot
d = datas[0]
source = ColumnDataSource(dict(x=d.x.values, y=d.y, z=d.z))
plot = figure()
plot.circle(x='x', y='y', size='z', source=source)

# Add a slide and it's update method
def update(attrname, old, new):
    n = int(slider.value)
    d = datas[n]
    source.data = dict(x=d.x, y=d.y, z=d.z)

slider = Slider(value=0, start=0, end=3, step=1)
slider.on_change('value', update)

curdoc().add_root(column(slider, plot))
```

Appreciate everyone for your help and for your excellent work of bokeh.

**Comment by gepcel** [comment]
After some testing, and javascript debugging under IE. Add a `Number.isInteger` method to `slider.js` will make sleders works in IE.  

```javascript
Number.isInteger = Number.isInteger || function(value) {
  return typeof value === 'number' && 
    isFinite(value) && 
    Math.floor(value) === value;
};
```

Some [reference doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger) here.

But I cann't figure out where should I modify in the source code. It seems like the file `C:\Anaconda3\Lib\site-packages\bokeh\server\static\js\bokeh-widgets.js:`, but no luck. And from another issue, the source code seems to be [here](https://github.com/bokeh/bokeh/blob/master/bokehjs/src/coffee/models/widgets/slider.coffee), but I cannot find this file locally. Where should I go from here?

**Comment by mattpap** [comment]
> `Number.isInteger`

We should have caught this earlier. IE is a big pain when it comes to modern browser APIs.

**Comment by mattpap** [comment]
> `ColumnDataSource's columns must be of the same length.`

It's important to have all columns of the same length and in bokeh 1.0 this will throw an exception instead. This is because for bokehjs it's a guessing game how to treat such data. There must be some obscure difference between IE and standards compliant browsers, which causes different behavior.

I'm not sure why `values` and `tolist()` give different results, and without more information (specifically type information) about the data frame and its series, I can only speculate about what's going on here. I would start from inspecting what `df.y.values` is compared to other series.

**Comment by bryevdv** [comment]
@gepcel agreed the `values` vs `tolist` this is obscure and I've never seen it before, it would be extremely helpful if you can share a minimum reproducing example code so we can investigate directly.

**Comment by gepcel** [comment]
@bryevdv @mattpap  I've managed to product a minimus example for the "inconsistent lengths" issue. How do you feel about openning another issue about the "inconsistent lengths", and leave this about "Number.isinteger()" of slider?

I'll wait before I upload the example with some description.

**Comment by bryevdv** [comment]
@gepcel that sounds like a good idea, thanks!

**Comment by bryevdv** [comment]
The PR did not actually fix the problem, the polyfil is not applied in IE:

<img width="551" alt="screen shot 2018-02-05 at 09 18 09" src="https://user-images.githubusercontent.com/1078448/35812036-88887d1e-0a55-11e8-8577-e9899038483e.png">

**Comment by bryevdv** [comment]
@mattpap can you comment? The polyfill code *is* run, it is just not useful:

<img width="563" alt="screen shot 2018-02-05 at 09 20 18" src="https://user-images.githubusercontent.com/1078448/35812169-e0d22704-0a55-11e8-9cca-00527b61b407.png">

**Comment by bryevdv** [comment]
This would seem to fix the problem, is this correct:
```
if (!Number_proto.isInteger) {
  Number.isInteger = function(value: number): boolean {
     return typeof value === 'number' && isFinite(value) && Math.floor(value) === value;
  }
 }
```

**Comment by mattpap** [comment]
> This (checking Number.prototype but setting on Number) would seem to fix the problem, is this correct:

No, `isInteger` is a class method (or something along this line), so one should check for `Number.isInteger` and assign to `Number.isInteger`.

**Comment by bryevdv** [comment]
confirmed fixed in `0.12.14rc2`
